### PR TITLE
feat: chat messages, model templates, layer zones

### PR DESCRIPTION
**Pull request by thiswillbeyourgithub**
- **minor**
- **minor: perf**
- **feat: support for chat templates**
- **use loguru instead of warnings**
- **feat: support for layer zones in addition to layer ids**
- **import make_dataset from utils instead of defining it in tests.py**
- **test: add a test for make_dataset**
- **test: update test values not passing**
- **doc: mention how to use chat templates**
- **doc: add a link related to OOM in transformers related to gguf**

Hi!

This is a polished version of #55 I made a while ago.

It mainly brings 3 features:
1. The hugging face model's template is respected if possible, making it easier to try a different model without having to dig into the intricacies of its template.
2. It is now possible to supply the examples in the form of familiar chat messages (list of dicts, with `role` and `content`).
3. It is now possible to specify `layer_zones` instead of `layer_ids`. For example `[[0.1, 0.5]]` means we control the layers whose relative depth is between 0.1 (included) and 0.5 (not included).

I took the liberty of adding [loguru](https://pypi.org/project/loguru/) for some logging but I can scrape that if you prefer.

All the tests are passing, although I had to add some value for the LLMs but they remain sound. I also added a test for the `make_dataset` function.

Hopefully, this will make `repeng` easier to work with when benchmarking my ideas over several setups and models.

Let me know if you want me to modify anything in this PR.

My plan is to now implement some mechanism to dump the logits to a file so that we can use commodity hardware, add more pairs of examples, use more advanced pairs manipulation (clustering etc).

Edit: my ongoing work will take place in [that fork](https://github.com/thiswillbeyourgithub/repeng-research-fork) and if I find other mistakes I'll push them here as #65 is not merged. I implemented a bunch of features, notably a h5py caching of activations to make it easier to run on low end hardware as this way I don't have to hold all the activations in memory.

If you could find the time to check out the README of the research fork and tell me what you think of the features I added I'm very interested :) I intend to make all my changes upstream so would like to know what you think of each!

### CODE REVIEW

**Comment by thiswillbeyourgithub** [comment]
Hi! Thanks for taking the time.

(note: there is a mistake in your message's formatting)

1. Interesting point about warnings.warn. I'm putting it back. Thanks.

2. The `continue_final_message` if False means that the prompt will have "end of message, now start a new message" added at the end of the prompt before asking the LLM to generate. So it depends on the use case. If you leave it to False, you can write the start of the assistant message yourself, for example starting with `Certainly, I will do as you said: ` will increase compliance rates.

3. Regarding make_dataset, there are tons of edge cases for example some models have not set their "tokenizer.chat_template" so I addressed this on my research fork. Also, not a fan of the non deterministic way to make a "dataset". In my fork I added a caching feature that stores the hidden activations for a given dataset and it works great to make quick runs. I think it's easy enough to create a "not tiny" dataset without having to use this kind of truncation technique.

Can't wait to get vgel's take on this to know what I should improve to make it upstream.

**Comment by wassname** [comment]
> online resources for me to grok this

Hmm it's been so long since I learned this stuff. Back in the day I read the Deep Learning Book (https://www.deeplearningbook.org/contents/prob.html) and did the cs231n course... but these days an LLM might be able to give you information that's more tailored to your background.

> My take is that the logprobs here are the last logit in the network, so taking the average gives us a measure of how surprised the model was by the words.

Well I'd say you get it, as this is pretty much right, imo.

Yeah, logprobs are often the last token. But in this case, we are taking the average logprobs of each token in the whole input sequence (e.g. the persona + suffix). So we are trying to measure, for a input sentence that the model reads, how likely would it be to generate it, on its own? If it's something it would never say, is it really worth "guiding" it not to do something.

> That's what I'm planning on buying in the coming months.

It's pretty good to be able to run 8b models, the ones smaller than that kind of suck.

**Comment by thiswillbeyourgithub** [comment]
> This looks good - I've been wanting to add something like layer_zones. Will pick at least that off and add you as co-author for 0.5. For chat templates I'll need to take a look at what you did closer - I'm leery of adding too much experiment-specific logic, but I guess having a "default" `make_dataset` in a utils file isn't bad as long as users can avoid it for custom things.

Thanks a lot. I really appreciate your thoughtfulness regarding crediting.

Regarding chat templates: It was so much of a pain to handle model specific quirks regarding templates etc that I ended up convinced that it makes sense to include it somewhere in repeng. A middle ground could be to put it in another file like `extra_utils.py` or `extras.py` so that users can use it if they want but remain aware that it's expected to be reworked if people only want specific models?

Re reading the code, your `train` expected as dataset a list of `DatasetEntry` which are strings. I see that I'm using `autocorrect_chat_templates` inside `read_representation` which is called by `train`. This was because I thought it was more logical to have `DatasetEntry` also be able to hold message chats as well as strings. Hence, and because `train` was expecting the `tokenizer` anyway, I converted from chat messages to strings via `autocorrect_chat_templates` directly inside `train`. An alternative could be to force `DatasetEntry` to only be a list of `str`, so that the conversion has to happen outside of `train`, and so autocorrect_chat_templates would appear more strikingly as "outside of the repeng architecture".

edit: of course let me know if I can be of help
